File: elixir/04.exs
defmodule Day04 do
    @moduledoc false

    defp read_input do
        "../inputs/input04.txt"
        |> Path.expand
        |> File.read!
        |> String.split("\n\n", trim: true)
        |> Enum.map(fn record ->
            record
            |> String.split(~r/\s/, trim: true)
            |> Enum.map(fn field ->
                field
                |> String.split(":")
            end)
        end)
    end

    @doc """
    Validate presence of required keys
    """
    def required_fields_present?(record) do
        required = ["byr", "iyr", "eyr", "hgt", "hcl", "ecl", "pid"]
        required
        |> Enum.all?(fn key ->
            record |> Enum.any?(fn pair ->
                Enum.at(pair, 0) == key
            end)
        end)
    end

    @doc """
    Validate year format
    """
    def valid_year?(val, min, max), do: val >= min && val <= max

    @doc """
    Validate hair colour format
    """
    def valid_hair_colour?(val), do: Regex.match?(~r/^#[0-9a-f]{6}$/, val)

    @doc """
    Validate height format
    """
    def valid_height?(val) do
        num = val |> String.replace(~r/(cm|in)?$/, "") |> String.to_integer
        unit = val |> String.replace(~r/^[0-9]+/, "")
        case unit do
            "cm" -> 150 <= num && num <= 193
            "in" -> 59 <= num && num <= 76
            ____ -> false
        end
    end

    @doc """
    Validate eye colour format
    """
    def valid_eye_colour?(val), do: val in ["amb", "blu", "brn", "gry", "grn", "hzl", "oth"]

    @doc """
    Validate pid format
    """
    def valid_pid?(val), do: Regex.match?(~r/^[0-9]{9}$/, val)

    @doc """
    Validate all aspects of a record
    """
    def fully_valid?(record) do
        cond do
            !required_fields_present?(record) ->
                false
            true ->
                record
                |> Enum.all?(fn field ->
                    [key, val] = field
                    case key do
                        "byr" -> valid_year?(String.to_integer(val), 1920, 2002)
                        "iyr" -> valid_year?(String.to_integer(val), 2010, 2020)
                        "eyr" -> valid_year?(String.to_integer(val), 2020, 2030)
                        "hcl" -> valid_hair_colour?(val)
                        "hgt" -> valid_height?(val)
                        "ecl" -> valid_eye_colour?(val)
                        "pid" -> valid_pid?(val)
                        _____ -> true
                    end
                end)
        end
    end

    def solve do
        # part 1 - 247
        read_input()
        |> Enum.count(&required_fields_present?/1)
        |> IO.puts

        # part 2 - 145
        read_input()
        |> Enum.count(&fully_valid?/1)
        |> IO.puts
    end
end

File: ruby/04.rb
input = File.read("../inputs/input04.txt")
@data = input.split("\n\n").map { |record|
    record.chomp.split(/\s/).map { |field|
        field.split(":")
    }
}

##
# Validate presence of required keys
# @param array
# @returns Boolean
def required_fields_present?(record)
    required = ["byr", "iyr", "eyr", "hgt", "hcl", "ecl", "pid"]
    required.all?{ |key|
        record.any?{ |pair|
            pair[0] == key
        }
    }
end

##
# Validate year format
# @param string val
# @returns Boolean
def valid_year?(val, min, max)
    val.length == 4 && val.to_i >= min && val.to_i <= max
end

##
# Validate height format
# @param string val
# @returns Boolean
def valid_height?(val)
    num, unit = val.match(/^([0-9]+)(cm|in)?$/)[1..2]
    num = num.to_i
    if unit == "cm"
        num >= 150 && num <= 193
    elsif unit == "in"
        num >= 59 && num <= 76
    else
        false
    end
end

##
# Validate hair colour format
# @param string val
# @returns Boolean
def valid_hair?(val)
    !(/^#[0-9a-f]{6}$/ =~ val).nil?
end

##
# Validate eye colour format
# @param string val
# @returns Boolean
def valid_eyes?(val)
    ["amb", "blu", "brn", "gry", "grn", "hzl", "oth"].include?(val)
end

##
# Validate pid format
# @param string val
# @returns Boolean
def valid_pid?(val)
    !(/^[0-9]{9}$/ =~ val).nil?
end

##
# Validate all aspects of a record
# @returns Boolean
def fully_valid?(record)
    return false if !required_fields_present? record
    record.each{ |field|
        case field[0]
        when "byr"
            return false if !valid_year?(field[1], 1920, 2002)
        when "iyr"
            return false if !valid_year?(field[1], 2010, 2020)
        when "eyr"
            return false if !valid_year?(field[1], 2020, 2030)
        when "hcl"
            return false if !valid_hair?(field[1])
        when "hgt"
            return false if !valid_height?(field[1])
        when "ecl"
            return false if !valid_eyes?(field[1])
        when "pid"
            return false if !valid_pid?(field[1])
        end
    }
    return true
end

# part 1 - 247
p @data.count{ |r| required_fields_present? r }

# part 2 - 145
p @data.count{ |r| fully_valid? r }
